internal/alb/sg/association.go

```diff
@@ -0,0 +1,259 @@
+package sg
+
+import (
+	"fmt"
+
+	"github.com/kubernetes-sigs/aws-alb-ingress-controller/internal/ingress/controller/store"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/service/ec2"
+	"github.com/kubernetes-sigs/aws-alb-ingress-controller/internal/alb/tg"
+	"github.com/kubernetes-sigs/aws-alb-ingress-controller/internal/aws/albec2"
+	"github.com/kubernetes-sigs/aws-alb-ingress-controller/internal/aws/albelbv2"
+	"github.com/kubernetes-sigs/aws-alb-ingress-controller/pkg/util/log"
+	"github.com/kubernetes-sigs/aws-alb-ingress-controller/pkg/util/types"
+)
+
+// Association represents the desired state of securityGroups & attachments for an Ingress resource.
+type Association struct {
+	// We identify Association by LbID
+	LbID string
+
+	LbArn          string
+	LbPorts        []int64
+	LbInboundCIDRs types.Cidrs
+	ExternalSGIDs  []string
+
+	Targets tg.TargetGroups
+}
+
+// AssociationController provides functionality to manage Association
+type AssociationController interface {
+	Reconcile(*Association) error
+	Delete(*Association) error
+}
+
+func NewAssociationController(store store.Storer, ec2 *albec2.EC2, elbv2 albelbv2.ELBV2API, logger *log.Logger) AssociationController {
+	lbAttachmentController := &lbAttachmentController{
+		elbv2:  elbv2,
+		ec2:    ec2,
+		logger: logger,
+	}
+	instanceAttachmentController := &instanceAttachmentController{
+		store:  store,
+		ec2:    ec2,
+		logger: logger,
+	}
+	sgController := &securityGroupController{
+		ec2:    ec2,
+		logger: logger,
+	}
+	namer := &namer{}
+	return &associationController{
+		lbAttachmentController:       lbAttachmentController,
+		instanceAttachmentController: instanceAttachmentController,
+		sgController:                 sgController,
+		namer:                        namer,
+		ec2:                          ec2,
+		logger:                       logger,
+	}
+}
+
+type associationController struct {
+	lbAttachmentController       LbAttachmentController
+	instanceAttachmentController InstanceAttachementController
+	sgController                 SecurityGroupController
+	namer                        Namer
+	ec2                          *albec2.EC2
+	logger                       *log.Logger
+}
+
+func (controller *associationController) Reconcile(association *Association) error {
+	if len(association.ExternalSGIDs) != 0 {
+		return controller.reconcileWithExternalSGs(association)
+	}
+	return controller.reconcileWithManagedSGs(association)
+}
+
+func (controller *associationController) Delete(association *Association) error {
+	return controller.deletedManagedSGs(association)
+}
+
+func (controller *associationController) reconcileWithExternalSGs(association *Association) error {
```

Comment:
It will be helpful to add some docs here to explain what is external SG. Similar to the managed SG. Or add a link if there is any

Reply:
added documentation to the code 😄

Reply:
external is probably just a bad name, they are 'externally managed' i guess? maybe we call the security groups managed and unmanaged?

Reply:
i'll change them to be externalManagedSGIDs 😄